### PR TITLE
 Intake | Docket Switch | Ability to manually close tasks

### PR DESCRIPTION
**Pull request by Sjones352**
Connects  #15099

### Description
Please explain the changes you made here.

### Acceptance Criteria
- [x] Please put this work behind the feature toggle: `docket_switch`
- [x] This feature should be accessible to the following user groups: `clerk of the board`
- [x] Once a user has added in information about the docket switch (date, type of grant, docket type), they should see the Add/Remove tasks screen
- [x] The progress bar should be on the second step (please see the designs below)
- [x] The user should see which docket they are switch from and to (For e.g: You are switching from Hearing to Direct Review.) at the top – this should be body text and bolded. 
- [x] The user should see which docket they are switch from and to (For e.g: You are switching from Hearing to Direct Review.) at the top – this should be body text and bolded. 
- [x] Below that, the user should see instructions – copy is in the designs below
- [x] The user should see a list of all orphan tasks (i.e. tasks NOT associated with a docket like FOIA, Translation etc) that are currently open or assigned on the current docket
  - [x] The orphan tasks should all be pre-selected
- [x] Tasks that would be automatically generated for the new docket (e.g. Distribution task) show in the list
  - [x] The automatically generated docket tasks that are applicable to the new docket should be inactive i.e. the user cannot remove it
- [ ] If a user re-selects the checkbox, that task should be transferred to the new docket
- [x] Selecting back will take the user back to the Review Request page
- [x] Taking action on this page is NOT required - the "continue" button should be active
- [ ] Include screenshot(s) in the Github issue if there are front-end changes

### User Facing Changes
 
Screen Shot 2021-01-15 at 12 58 50 AM

### Storybook Story
*For Frontend (Presentationa) Components*
* [x] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

### CODE REVIEW

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx [hunk 1]
```diff
@@ -17,13 +19,26 @@ export const DocketSwitchReviewRequestContainer = () => {
...
+      external_id: appeal.externalId,
+      assigned_to_type: 'User',
+      docket_name: appeal.docketName,
+      reciept_date: formData.receiptDate,
+      issues: appeal.issues,
+      appeallant_name: appeal.appellantFullName
+    };
+    
+    try {
+      await dispatch(completeDocketSwitchGranted(data));
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx:36]: this should be the `updatetDockettSwitch` action — we aren't submitting to the backend until after the final step

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx [hunk 2]
```diff
@@ -17,13 +19,26 @@ export const DocketSwitchReviewRequestContainer = () => {
...
   const handleCancel = () => goBack();
 
-  const handleSubmit = () => {
-    // Add stuff to redux store
+  const handleSubmit = async (formData) => {
 
-    // Call stepForward redux action
-    dispatch(stepForward());
+    const data = {
+      parent_id: taskId,
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx:25]: All that we need to do here is store the couple of things from this form to Redux. We don't need to worry about creating tasks — that will be handled on the backend after the final step is complete

So basically, this should be something like this:

```js
const data = {
formData: {
receiptDate: formData.receiptDate,
docketType: formData.docketType,
}
}
```

File: client/app/queue/docketSwitch/grant/docketSwitchGrantedSlice.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import ApiUtil from '../../../util/ApiUtil';
+
+export const completeDocketSwitchGranted = createAsyncThunk(
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/docketSwitchGrantedSlice.js:4]: While we will eventually need a thunk for submitting the docket switch stuff to the backend, that will come at the end of the process. And we'll probably want to just put the submission thunk in the existing `docketSwitchSlice.js` file.

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  const { goBack, push } = useHistory();
+
+  const appeal = useSelector((state) =>
+    appealWithDetailSelector(state, { appealId })
+  );
+
+  const tasks = useSelector((state) =>
+    getAllTasksForAppeal(state, { appealId })
+  );
+
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskContainer.jsx:19]: You'll want to pull out any info from Redux that you'll need from the previous step:

const { docketType } = useSelector((state) => state.docketSwitch.formData);

File: client/app/queue/docketSwitch/docketSwitchSlice.js [hunk 5]
```diff
@@ -26,14 +34,31 @@ const docketSwitchSlice = createSlice({
         ...state.formData,
         ...updates,
```
Comment (jcq) [inline client/app/queue/docketSwitch/docketSwitchSlice.js:35]: So, there are a couple of things that have to change here (my bad from original setup). The contents of `updateDocketSwitch` should be this instead:

```js
const { formData: updates } = action.payload;

      // eslint-disable-next-line no-undefined
      updates.receiptDate = updates.receiptDate?.toISOString() ?? undefined;

      state.formData = {
        ...state.formData,
        ...updates,
      };

```

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  const docketType = useSelector((state) => 
+    state.docketSwitch.formData.formData.docketType);
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskContainer.jsx:22]: This has an extra `formData`; it should be this instead:
```js
const docketType = useSelector((state) => 
    state.docketSwitch.formData.docketType);
```

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx [hunk 7]
```diff
@@ -17,13 +17,24 @@ export const DocketSwitchReviewRequestContainer = () => {
...
 
-  const handleSubmit = () => {
-    // Add stuff to redux store
+  const handleSubmit = async (formData) => {
 
-    // Call stepForward redux action
-    dispatch(stepForward());
-
-    // Move to next step
+    const data = {
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx:22]: After thinking about it, we can actually remove this `data` variable altogether and use as noted below

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx [hunk 8]
```diff
@@ -17,13 +17,24 @@ export const DocketSwitchReviewRequestContainer = () => {
...
+      formData: {
+      disposition: formData.disposition,
+      docketType: formData.docketType,
+      receiptDate: formData.receiptDate,
+      issueIds: formData.issueIds
+     }
+    }
+    
+    try {
+      await dispatch(updateDocketSwitch(data));
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestContainer.jsx:32]: The `formData` we already have is all we actually need, so we can simplify things. Can just do this instead:
```js
await dispatch(updateDocketSwitch({data}));
```

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  DOCKET_SWITCH_GRANTED_MODAL_INSTRUCTION
+} from 'app/../COPY';
+import { sprintf } from 'sprintf-js';
+import { yupResolver } from '@hookform/resolvers';
+import * as yup from 'yup';
+import { css } from 'glamor';
+import ReactMarkdown from 'react-markdown';
+import CheckboxGroup from 'app/components/CheckboxGroup';
+import Button from '../../../components/Button';
+import _ from 'lodash';
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx:22]: Not sure if you need `omitBy`, but definitely don't want to import the entirety of `lodash`

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+  const sectionStyle = css({ marginBottom: '24px' });
+
+  const taskOptions = useMemo(() => {
+    return taskListing && taskListing.map((task) =>
+      ({ label: task.label,
+        id: task.taskId.toString() }));
+  }, [taskListing]);
+
+  const selectedIssues = _.omitBy(tasks, false);
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx:54]: Seems like we don't need this...?

**Review by jcq — COMMENTED** [review]
This is looking good. After talking to Sally, we decided to move the part about the mandatory tasks from this ticket/PR to mine. That means you don't need to worry about the part on the design with the disabled checkbox (for "Distribution Task" in the design).

So it looks like the only thing that's still missing is a Jest test. I would say don't worry about going too deep into that, as I'll need to be adding to the same thing in my follow-up PR. Maybe just create the file and do a simple snapshot test?

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+                options={taskOptions}
+                onChange={(event) => onCheckChange(handleTaskChange(event))}
+                styling={css({ marginBottom: '0' })}
+                values={tasks}
+              />
+            );
+          }}
+        />
+
+        { activeTaskId && (
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx:129]: Currently, the modal is showing up both for removing and for re-checking a task. My interpretation of the AC is that it'd only be for unchecking the task.

You'll want to check if the active task is changing to checked or unchecked (maybe in the onCheckChange or handleTaskChange functions). 
If changing to unchecked, that's when we want to show the modal.
If changing to checked, you can go straight to updating the task selections during the onChange for the checkbox.

**Review by leikkisa — COMMENTED** [review]
Hi Sandra - really thorough, amazing PR. It's cool that the list of tasks is connected too.

JC and I discussed and we think it makes sense to move the "tasks that are created" to his PR after checking in with Rutvi, so removing that from the AC for this one.

I think there's one thing that needs to be addressed, which is to not show the "Removal confirmation" modal when re-checking a task. Otherwise looks great to me!

**Comment by leikkisa** [comment]
PS - the flakey test is my fault - being fixed in another PR.

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.stories.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  component: DocketSwitchAddTaskForm,
+  decorators: [
+    // AppSegment styling relies on being inside a .cf-content-inside element
+    (storyFn) => <div className="cf-content-inside">{storyFn()}</div>,
+  ],
+  parameters: {},
+  args: {
+    appellantName: 'Jane Doe',
+    taskListing: tasks,
+    docketName: 'Direct Review',
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.stories.js:21]: I think this needs to be renamed to `docketFrom`

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.stories.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    docketType: 'Hearings',
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.stories.js:22]: ...and this needs to be `docketTo`

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    const newValues = {};
+
+    taskListing.forEach((item) => (newValues[item.taskId] = true));
+    setTasks(newValues);
+
+    setValue('taskIds', taskListing.map((task) => task.taskId));
+  };
+
+  const activeTaskLabel = useMemo(() => {
+    return activeTaskId ? taskListing.find((task) => task.taskId === activeTaskId)?.['label'] : null;
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx:63]: To ensure that our `===` operator works, let's cast the (potentially numeric) ID values to strings:

```js
const activeTaskLabel = useMemo(() => {
    return activeTaskId ?
      taskListing.find(
        (task) => String(task.taskId) === String(activeTaskId)
      )?.['label'] :
      null;
  }, [activeTaskId]);
```

**Review by jcq — APPROVED** [review]
A couple of final things to change (see comments), but looking good!

**Comment by leikkisa** [comment]
Also, here's some code changes in `DocketSwitchAddTaskForm` that prevent the modal from showing when re-checking a task:
```ruby
  const handleTaskChange = (evt) => {
    const targetTaskId = evt.target.id.toString();
    setActiveTaskId(targetTaskId);

    if (!tasks[targetTaskId] === true) {
      updateTaskSelections(targetTaskId);
    };
  };


  const updateTaskSelections = (targetTaskId = null) => {
    const updatedTaskId = activeTaskId || targetTaskId;
    const newSelections = {
      ...tasks,
      [updatedTaskId]: !tasks[updatedTaskId],
    };

    setTasks(newSelections);
    setActiveTaskId(null);
    setValue(
      'taskIds',
      Object.keys(newSelections).filter((key) => newSelections[key])
    );
  };
```

File: client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                styling={css({ marginBottom: '0' })}
+                values={tasks}
+              />
+            );
+          }}
+        />
+
+        { activeTaskId && (
+          <DocketSwitchAddTaskModal
+            onSubmit={onSubmit}
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddTaskForm.jsx:131]: Also, I don't think this line is necessary or doing anything.

**Pull request merged**